File: server/src/archives/metadata/koromo.rs
use super::{ArchiveMetadata, MultiTextField};
use crate::{archives, config::CONFIG, utils};
use serde::Deserialize;

#[derive(Deserialize)]
pub struct Metadata {
  #[serde(rename = "Title")]
  pub title: String,
  #[serde(rename = "Description", default)]
  pub description: Option<String>,
  #[serde(rename = "Source", default)]
  pub source: Option<String>,
  #[serde(rename = "URL", default)]
  pub url: Option<String>,
  #[serde(rename = "Artist", default)]
  pub artists: Option<MultiTextField>,
  #[serde(rename = "Groups", alias = "Circle", default)]
  pub groups: Option<MultiTextField>,
  #[serde(rename = "Magazine", default)]
  pub magazines: Option<MultiTextField>,
  #[serde(rename = "Publisher", default)]
  pub publishers: Option<MultiTextField>,
  #[serde(rename = "Parody", default)]
  pub parodies: Option<MultiTextField>,
  #[serde(rename = "Tags", default)]
  pub tags: Option<Vec<String>>,
  #[serde(rename = "Thumbnail", default)]
  pub thumb_index: Option<i16>,
  #[serde(rename = "Released", default)]
  pub released: Option<i64>,
}

pub fn add_metadata(info: Metadata) -> ArchiveMetadata {
  let mut archive = ArchiveMetadata {
    title: if CONFIG.metadata.parse_filename_title {
      utils::parse_filename(&info.title).0.unwrap_or(info.title)
    } else {
      info.title
    },
    description: info.description,
    thumbnail: info.thumb_index.map(|index| index + 1),
    released_at: utils::map_timestamp(info.released),
    artists: info.artists.map(|field| field.to_vec()),
    circles: info.groups.map(|field| field.to_vec()),
    magazines: info.magazines.map(|field| field.to_vec()),
    publishers: info.publishers.map(|field| field.to_vec()),
    parodies: info.parodies.map(|field| field.to_vec()),
    tags: info.tags.map(|tags| {
      tags
        .into_iter()
        .map(|tag| (utils::capitalize_words(&tag), "".to_string()))
        .collect()
    }),
    ..Default::default()
  };

  let mut sources: Vec<archives::Source> = vec![];

  if let Some(url) = info.url {
    sources.push(archives::Source {
      name: utils::parse_source_name(&url),
      url: Some(url),
    });
  }

  if let Some(url) = info.source {
    sources.push(archives::Source {
      name: utils::parse_source_name(&url),
      url: Some(url),
    });
  }

  if !sources.is_empty() {
    archive.sources = Some(sources);
  }

  archive
}

File: cli/metadata/koromo.ts
import dayjs from 'dayjs';
import { z } from 'zod';
import config from '../../shared/config';
import { type ArchiveMetadata } from '../../shared/metadata';
import { mapMultiField, multiTextField } from './schemas';
import { parseFilename } from './utils';

const metadataSchema = z.object({
	Title: z.string(),
	Description: z.string().nullable().optional(),
	Source: z.string().nullable().optional(),
	URL: multiTextField.nullable().optional(),
	Artists: multiTextField.nullable().optional(),
	Artist: multiTextField.nullable().optional(),
	Circle: multiTextField.nullable().optional(),
	Groups: multiTextField.nullable().optional(),
	Magazine: multiTextField.nullable().optional(),
	Magazines: multiTextField.nullable().optional(),
	Publisher: multiTextField.nullable().optional(),
	Publishers: multiTextField.nullable().optional(),
	Event: multiTextField.nullable().optional(),
	Events: multiTextField.nullable().optional(),
	Parody: multiTextField.nullable().optional(),
	Parodies: multiTextField.nullable().optional(),
	Series: multiTextField.nullable().optional(),
	Characters: multiTextField.nullable().optional(),
	Tags: multiTextField.nullable().optional(),
	Thumbnail: z.number().nullable().optional(),
	Released: z.number().nullable().optional(),
	Published: z.number().nullable().optional(),
});

export default async (content: string, archive: ArchiveMetadata) => {
	const parsed = JSON.parse(content);
	const { data, error } = metadataSchema.safeParse(parsed);

	if (!data) {
		throw new Error(`Failed to parse Koromo metadata: ${error}`);
	}

	archive = structuredClone(archive);

	if (config.metadata?.parseFilenameAsTitle) {
		archive.title = parseFilename(data.Title)[0] ?? data.Title;
	} else {
		archive.title = data.Title;
	}

	archive.description = data.Description ?? undefined;
	archive.thumbnail =
		data.Thumbnail !== null && data.Thumbnail !== undefined ? data.Thumbnail + 1 : undefined;
	archive.releasedAt = (() => {
		if (data.Released) {
			return dayjs.unix(data.Released).toDate();
		}

		if (data.Published) {
			return dayjs.unix(data.Published).toDate();
		}

		return undefined;
	})();

	archive.tags = [];

	for (const tag of [...mapMultiField(data.Artist), ...mapMultiField(data.Artists)]) {
		archive.tags.push({ namespace: 'artist', name: tag });
	}

	for (const tag of [...mapMultiField(data.Circle), ...mapMultiField(data.Groups)]) {
		archive.tags.push({ namespace: 'circle', name: tag });
	}

	for (const tag of [...mapMultiField(data.Magazine), ...mapMultiField(data.Magazines)]) {
		archive.tags.push({ namespace: 'magazine', name: tag });
	}

	for (const tag of [...mapMultiField(data.Event), ...mapMultiField(data.Events)]) {
		archive.tags.push({ namespace: 'event', name: tag });
	}

	for (const tag of [...mapMultiField(data.Publisher), ...mapMultiField(data.Publishers)]) {
		archive.tags.push({ namespace: 'publisher', name: tag });
	}

	for (const tag of [
		...mapMultiField(data.Parody),
		...mapMultiField(data.Parodies),
		...mapMultiField(data.Series),
	]) {
		archive.tags.push({ namespace: 'parody', name: tag });
	}

	for (const tag of mapMultiField(data.Tags)) {
		const [namespace, name] = tag.split(':');
		if (namespace && name) {
			archive.tags.push({ namespace, name });
		} else {
			archive.tags.push({ namespace: 'tag', name: tag });
		}
	}

	for (const tag of mapMultiField(data.Characters)) {
		archive.tags.push({ namespace: 'character', name: tag });
	}

	archive.sources = [];

	if (data.URL?.length) {
		for (const url of mapMultiField(data.URL)) {
			archive.sources.push({ url });
		}
	}

	if (data.Source) {
		archive.sources.push({ url: data.Source });
	}

	return archive;
};
